### PR TITLE
Pre-Submit Tryjobs for Flutter Gold

### LINKED ISSUE #43995: Better Triage Notification System for Gold
The [Flutter Gold dashboard](https://flutter-gold.skia.org/?query=source_type%3Dflutter) does not currently support notifications. 
This is a sticking point in the case where:
- Pull request 1 lands, changing pre-existing golden file A
- Triage is not completed to approve the change
- The change ignore expires
- Pull request 2 picks up change to golden file A from master
- Pre-submit checks for PR 2 fail since there is no approved match for updated golden file A

This is not an often occurring case, as golden files are not changed often (they are visual constants!), and triage breakdown is also uncommon with considerate ignore practice.

To address this case, I would like to implement a notification system, likely through [flutter/cocoon](https://github.com/flutter/cocoon), that will ping PRs or directly email ignore owners before they expire, closing the potential gap for triage breakdown. This notification would encourage the owner to either 
- triage images (if change has landed)
- extend ignore (if change has not landed)
- delete ignore (if triage has been completed, or if the change is not actually ready to land)

### LINKED ISSUE #45181: Enable Tryjobs for Flutter Gold
Gold can now support tryjobs from github! 
The FlutterPreSubmitFileComparator will need to be updated to make use of this new feature. :)

Ref: https://skia-review.googlesource.com/c/buildbot/+/253698

### PR DESCRIPTION
**Pull request by Piinks**
## Description

This PR adds support for pre-submit tryjobs for Flutter Gold. This will allow for golden file tests to be run against baselines in pre-sbumit tests, generating visual diffs and enabling pre-submit triage of new or changed golden files.

**Status:**
- [x] https://skia-review.googlesource.com/c/buildbot/+/256018 + roll to CIPD needed for patchset_id
- [x] Update wiki

## Related Issues

- Fixes #43995 somewhat in that triage will be handled before a change lands, removing the need to remember to visit the dashboard and triage.
- Fixes #45181

## Tests

Updated related tests.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by Piinks** [comment]
Blocked by #44479 :)

File: packages/flutter_goldens_client/lib/skia_client.dart [hunk 1]
```diff
@@ -193,6 +193,81 @@ class SkiaGoldClient {
...
+        .path,
+      '--commit', commitHash,
+      '--keys-file', keys.path,
+      '--failure-file', failures.path,
+      '--passfail',
+      '--crs', 'github',
+      '--changelist', pullRequest,
+      '--cis', 'cirrus',
+      '--jobid', cirrusTaskID,
+      '--patchset', nthCommit,
```
Comment (Piinks) [inline packages/flutter_goldens_client/lib/skia_client.dart]: @kjlubick, is this right? Or can I cut some of these arguments for the tryjob?

File: packages/flutter_goldens_client/lib/skia_client.dart [hunk 2]
```diff
@@ -362,6 +437,19 @@ class SkiaGoldClient {
...
+  /// Returns the number of commits made on the current branch.
+  Future<String> getBranchCommitCount() async {
+    final String branch = platform.environment['CIRRUS_BRANCH'];
+    final String masterBranch = platform.environment['CIRRUS_BASE_BRANCH'];
+
+    // This git wizardry does not appear to work as intended on Cirrus.
+    final io.ProcessResult revList = await process.run(
+      <String>['git', 'rev-list', branch, '^$masterBranch', '--count'],
+      workingDirectory: _flutterRoot.path,
+    );
```
Comment (Piinks) [inline packages/flutter_goldens_client/lib/skia_client.dart]: @kjlubick, getting the nth commit like I planned to doesn't work on Cirrus, I'm looking into another way to get the number.

File: packages/flutter_goldens/lib/flutter_goldens.dart [hunk 3]
```diff
@@ -299,40 +299,18 @@ class FlutterPreSubmitFileComparator extends FlutterGoldenFileComparator {
...
   @override
   Future<bool> compare(Uint8List imageBytes, Uri golden) async {
     golden = _addPrefix(golden);
-    final String testName = skiaClient.cleanTestName(golden.path);
-    final List<String> testExpectations = skiaClient.expectations[testName];
-
-    if (testExpectations == null) {
-      // There is no baseline for this test
-      return true;
-    }
```
Comment (Piinks) [inline packages/flutter_goldens/lib/flutter_goldens.dart]: @kjlubick, what result will be returned if the tryjob does not have a baseline?

Comment (Piinks) [reply packages/flutter_goldens/lib/flutter_goldens.dart]: Oh, I forgot. I am guessing it will return false, and then be triaged.

**Comment by Piinks** [comment]
I'll have to make a framework change to actually trigger a framework test and see that this works... looking for a typo or something similar...

File: packages/flutter/lib/src/animation/animation_controller.dart [hunk 4]
```diff
@@ -41,13 +41,14 @@ const Tolerance _kFlingTolerance = Tolerance(
...
+/// disabled.
 ///
 /// When [AccessibilityFeatures.disableAnimations] is true, the device is asking
 /// Flutter to reduce or disable animations as much as possible. To honor this,
-/// we reduce the duration and the corresponding number of frames for animations.
-/// This enum is used to allow certain [AnimationController]s to opt out of this
-/// behavior.
+/// we reduce the duration and the corresponding number of frames for
+/// animations. This enum is used to allow certain [AnimationController]s to opt
+/// out of this behavior.
```
Comment (Piinks) [inline packages/flutter/lib/src/animation/animation_controller.dart]: This change is a trivial 80 character line limit correction in keeping with the style guide, and to also trigger the relevant framework tests confirming tryjob execution.

File: packages/flutter_goldens_client/lib/skia_client.dart (hunk 1 above)
Comment (kjlubick) [reply packages/flutter_goldens_client/lib/skia_client.dart]: Lines 222-226 lgtm for tryjobs; The others also look fine and are needed for all such `imgtest init` invocations

File: packages/flutter_goldens_client/lib/skia_client.dart (hunk 2 above)
Comment (kjlubick) [reply packages/flutter_goldens_client/lib/skia_client.dart]: That's fine; To test things out you could hard-code a number (e.g. 1) for now

File: packages/flutter_goldens_client/lib/skia_client.dart (hunk 1 above)
Comment (Piinks) [reply packages/flutter_goldens_client/lib/skia_client.dart]: Right now, this is failing:
```
Exit code 1: Skia Gold tryjobInit failure.
stdout: 
stderr: Error: unknown flag: --crs
```

Comment (kjlubick) [reply packages/flutter_goldens_client/lib/skia_client.dart]: Remind me where you get goldctl?  If it's from CIPD, I probably need to update it.

File: packages/flutter_goldens/lib/flutter_goldens.dart (hunk 3 above)
Comment (kjlubick) [reply packages/flutter_goldens/lib/flutter_goldens.dart]: Yeah, that should be correct

File: packages/flutter_goldens_client/lib/skia_client.dart (hunk 1 above)
Comment (Piinks) [reply packages/flutter_goldens_client/lib/skia_client.dart]: Ah! Yes, we get goldctl from CIPD.

**Comment by Piinks** [comment]
It looks like this is a bit flaky. 
On the failing test shards, some tests have failed with this exception from 

`goldctl imgtest add ...`


```
Exit code 1: Skia Gold tryjobAdd failure.
stdout:
stderr: Error running command: ''gold client not ready: Gold results fields invalid: Received nil
pointer for GoldResult. At jsonio.go:194 goldclient.go:452 goldclient.go:289 goldclient.go:279
cmd_imgtest.go:334 command.go:833 command.go:917 main.go:45 proc.go:203 asm_amd64.s:1357. At
goldclient.go:453 goldclient.go:289 goldclient.go:279 cmd_imgtest.go:334 command.go:833
command.go:917 main.go:45 proc.go:203 asm_amd64.s:1357''

When the exception was thrown, this was the stack:
#0      SkiaGoldClient.tryjobAdd (package:flutter_goldens_client/skia_client.dart:284:7)
<asynchronous suspension>
#1      FlutterPreSubmitFileComparator.compare (package:flutter_goldens/flutter_goldens.dart:313:23)
<asynchronous suspension>
```

The tryjob did show up on the dashboard here: https://flutter-gold.skia.org/changelists?page_size=50
Clicking on it results in a 404 error.

File: packages/flutter_goldens_client/lib/skia_client.dart (hunk 1 above)
Comment (kjlubick) [reply packages/flutter_goldens_client/lib/skia_client.dart]: Updated now in CIPD

**Comment by kjlubick** [comment]
Around 22:07 UTC (last night), I see from the ingestion logs that files were uploaded to the right place and processed.
I see some entries in Firestore:
 - the ChangeList was made, and a Patchset 1 under it
 - 6 cirrus TryJobs
 - oodles of TryJob results

From that conclusion, data is (mostly) getting into Gold successfully.

**Comment by kjlubick** [comment]
Ah, the data in Firestore is slightly wrong - the TryJob and TryJobResults are incorrectly tagged as being "gerrit"

**Comment by kjlubick** [comment]
Fixes for the gold-side of things are in https://skia-review.googlesource.com/c/buildbot/+/255778

Not sure yet about that nil dereference - Are you running multiple tests at the same time?  goldctl is not thread safe

File: packages/flutter_goldens_client/lib/skia_client.dart [hunk 5]
```diff
@@ -193,6 +193,99 @@ class SkiaGoldClient {
...
+      '--png-file', goldenFile.path,
+    ];
+
+    final io.ProcessResult result = await io.Process.run(
+      _goldctl,
+      imgtestArguments,
+    );
+
+    if (result.exitCode != 0) {
+      final StringBuffer buf = StringBuffer()
```
Comment (kjlubick) [inline packages/flutter_goldens_client/lib/skia_client.dart]: Can you maybe log the arguments if the exit code is non zero?

Comment (kjlubick) [reply packages/flutter_goldens_client/lib/skia_client.dart]: As extra debugging (for the flaky failures), it would be nice to know what the contents of the result-state.json file in that working directory are.

File: packages/flutter_goldens_client/lib/skia_client.dart [hunk 6]
```diff
@@ -193,6 +193,99 @@ class SkiaGoldClient {
...
+  /// result for the tryjob.
+  ///
+  /// The testName and goldenFile parameters reference the current comparison
+  /// being evaluated by the [FlutterSkiaGoldFileComparator].
+  Future<bool> tryjobAdd(String testName, File goldenFile) async {
+    assert(testName != null);
+    assert(goldenFile != null);
+
+    final List<String> imgtestArguments = <String>[
+      'imgtest', 'add',
```
Comment (kjlubick) [inline packages/flutter_goldens_client/lib/skia_client.dart]: arguments look good

**Comment by Piinks** [comment]
> Not sure yet about that nil dereference - Are you running multiple tests at the same time? goldctl is not thread safe

AFAIK, for each test shard (i.e. `framework_tests-libraries-windows`, `framework_tests-libraries-mac`) we fork processes for each test file, but they don't run concurrently. 
`goldctl imgtest init` is called for each test file, followed by the `goldctl imgtest add` for each test therein. 

The actual shards may run at the same time on Cirrus though. 🤔 

Thanks for investigating! I have pushed up a change that will provide more debugging information, including the contents of result-state.json. It looks like `init` is where it is flaking out right now.

**Comment by kjlubick** [comment]
Multiple shards using goldctl at the same time from different machines == fine
one machine using goldctl from multiple threads in the same work directory == not fine
one machine using goldctl from multiple threads in different work directories == probably fine

Hopefully the debugging tells us more

**Comment by Piinks** [comment]
Looking lovely now! 😄  I'll be moving forward with getting this ready to land.

Instead of nth commit for the patch set, would it be ok to just use the current commit? I'm still looking into ways to get the nth commit, but received an offline suggestion to just go with the hash.

File: packages/flutter_goldens_client/lib/skia_client.dart [hunk 7]
```diff
@@ -223,12 +222,12 @@ class SkiaGoldClient {
       '--changelist', pullRequest,
       '--cis', 'cirrus',
       '--jobid', cirrusTaskID,
-      '--patchset', nthCommit,
+      '--patchset', commitHash,
```
Comment (kjlubick) [inline packages/flutter_goldens_client/lib/skia_client.dart]: This will make ingestion sad, I believe. However, I can update gold to take the PatchSet ID (i.e. the commit hash) and turn it into the nth commit on our side.  This will also likely require a push to goldctl and a slightly different flag (--patchset_id)

**Review by kjlubick — CHANGES_REQUESTED** [review]

File: packages/flutter_goldens_client/lib/skia_client.dart (hunk 7 above)
Comment (kjlubick) [reply packages/flutter_goldens_client/lib/skia_client.dart]: https://skia-review.googlesource.com/c/buildbot/+/256018 will have that (plus a CL to roll goldctl in CIPD)

Comment (Piinks) [reply packages/flutter_goldens_client/lib/skia_client.dart]: Ah I see. Thank you for adding that option - and so quickly! :+1:

Comment (kjlubick) [reply packages/flutter_goldens_client/lib/skia_client.dart]: CIPD has the new version and ingestion servers should be able to take it

File: packages/flutter_goldens/lib/flutter_goldens.dart [hunk 8]
```diff
@@ -131,15 +131,15 @@ abstract class FlutterGoldenFileComparator extends GoldenFileComparator {
   /// Calculate the appropriate basedir for the current test context.
```
Comment (kjlubick) [inline packages/flutter_goldens/lib/flutter_goldens.dart]: Driveby comment: you may want to add to the documentation here explaining what suffix is and why it is needed.

File: packages/flutter_goldens_client/lib/skia_client.dart [hunk 9]
```diff
@@ -209,6 +209,96 @@ class SkiaGoldClient {
     return true;
   }
 
+  /// Executes the `imgtest init` command in the goldctl tool for tryjobs.
+  ///
+  /// The `imgtest` command collects and uploads test results to the Skia Gold
+  /// backend, the `init` argument initializes the current tryjob.
+  Future<void> tryjobInit() async {
+    final File keys = workDirectory.childFile('keys.json');
+    final File failures = workDirectory.childFile('failures.json');
```
[the remark this answers is not in the document]
Comment (dnfield) [reply packages/flutter_goldens_client/lib/skia_client.dart]: If I'm reading this right, it's coming from https://github.com/flutter/flutter/pull/44474/files#diff-259a52e89a62c0e0633f88f4307c22b7L142, which is not tracked by source control - is that right?

File: packages/flutter_goldens/lib/flutter_goldens.dart [hunk 10]
```diff
@@ -129,6 +129,11 @@ abstract class FlutterGoldenFileComparator extends GoldenFileComparator {
   Uri getTestUri(Uri key, int version) => key;
 
   /// Calculate the appropriate basedir for the current test context.
+  ///
+  /// The optional [suffix] argument is used by the
+  /// [FlutterSkiaGoldFileComparator] and the [FlutterPreSubmitFileComparator].
+  /// These [FlutterGoldenFileComparators] randomize their base directories to
+  /// maintain thread safety while using the `goldctl` tool.
```
Comment (dnfield) [inline packages/flutter_goldens/lib/flutter_goldens.dart]: Why aren't we just using `createTempSync` from package:file?

https://pub.dev/documentation/file/latest/file/Directory/createTempSync.html

File: packages/flutter_goldens_client/lib/skia_client.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/flutter_goldens_client/lib/skia_client.dart]: > Is workDirectory in source control? 

I am not sure what you mean by this.

I thought it was automatically blown away on ci since we start with a clean instance for each build. I can split it out further for ci/non-ci code paths, but I thought this would keep it from becoming a more complicated code path for those cases.

File: packages/flutter_goldens/lib/flutter_goldens.dart (hunk 10 above)
Comment (Piinks) [reply packages/flutter_goldens/lib/flutter_goldens.dart]: When run locally, we want the files to be accessible to see diffs etc.

Comment (dnfield) [reply packages/flutter_goldens/lib/flutter_goldens.dart]: Ahh.  I'm a bit curious about whether we'd hit collisions at some point using math.random, but it's probably fine.

File: packages/flutter_goldens_client/lib/skia_client.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (dnfield) [reply packages/flutter_goldens_client/lib/skia_client.dart]: Yeah this should be fine.  I was worried we were modifying something that the Flutter source control cared about with this.

**Review by dnfield — APPROVED** [review]
LGTM

File: packages/flutter_goldens/lib/flutter_goldens.dart (hunk 10 above)
Comment (Piinks) [reply packages/flutter_goldens/lib/flutter_goldens.dart]: I thought so too, but then realized we only run 2-3 processes concurrently on ci, and only ~130/~4500 tests are golden file tests, so I feel like a collision would be incredibly rare in how this is written.

**Comment by Piinks** [comment]
Thanks for the review @dnfield! I just need to update the wiki and this will be ready to go.

**Comment by Piinks** [comment]
> This pull request is not suitable for automatic merging in its current state.
> 
> * This pull request has changes requested. Please resolve those before re-applying the label.


@dnfield I haven't seen this before. What does it mean?

**Comment by dnfield** [comment]
It's a bug :( It'ssupposed to be checking for reviews and refuse to merge if there are open change requests.  This PR doesn't have any.  Unfortunately the bot won't merge it currently.

**Comment by Piinks** [comment]
> It's a bug :( It'ssupposed to be checking for reviews and refuse to merge if there are open change requests. This PR doesn't have any. Unfortunately the bot won't merge it currently.

No worries! I was curious. :)

**Comment by skia-gold** [comment]
test message from gold

**Comment by Piinks** [comment]
> test message from gold

I read you gold leader. ⭐️ 😛
